module/zfs/dmu.c

```diff
@@ -772,6 +773,44 @@ dmu_objset_zfs_unmounting(objset_t *os)
 	return (B_FALSE);
 }
 
+/*
+ * Return the amount of data contained in the given dnode's chunk
+ */
+static uint64_t
+dmu_dnode_count_data_in_chunk(dnode_t *dn, uint64_t offset, uint64_t chunk_len)
+{
+
+	int err;
+	uint64_t data_size = chunk_len;
+	uint64_t hole_end_off, hole_off = offset;
+
+	while (hole_off < offset + chunk_len) {
+		/*
+		 * Locate next hole offset. If there is an error or our
+		 * next hole is beyond the chunk, we're done.
+		 */
+		err = dnode_next_offset(dn,
+		    DNODE_FIND_HOLE, &hole_off, 1, 1, 0);
+		if (err != 0 || hole_off > offset + chunk_len)
+			break;
+
+		/*
+		 * Locate the hole end. If there is an error or our
+		 * offset is beyond the end of the chunk we round down
+		 * to the end of our chunk.
+		 */
+		hole_end_off = hole_off;
+		err = dnode_next_offset(dn, 0, &hole_end_off, 1, 1, 0);
+		if (err != 0 || hole_end_off > offset + chunk_len)
+			hole_end_off = offset + chunk_len;
+
+		data_size -= hole_end_off - hole_off;
```

Comment:
nit: personally I'd suggest not using `-=` here and expanding this to use just `=`, but I'm ok with it either way.